[issue-65] Rework FlinkPravegaInputFormat to use batch client

tzulitai opened the pull request:
**Change log description**

- Reworks the `FlinkPravegaInputFormat` to use the experimental Pravega batch client.
- Expand current IT tests for the input format. The tests now additionally cover the following verifications: 1) reading multiple streams with batch, 2) multiple passes, and 3) rewinds input format in case of failures.
- Previous configuration of the input format such as `startReadTimestamp` and `eventReadTimeout` is removed, since they are no longer relevant / useful after switching to the batch client.

**Purpose of the change**

This PR closes #65 and #56.

As discussed in #65 and #56, the use of reader groups / the `EventStreamReader` API does not work well in batch scenarios, which only allows Pravega segments to be read sequentially. There's also the problem that residual reader group states does not (gracefully) allow the input format to have multiple passes, or even re-executions on job failures.

This PR reworks the input format to use the experimental batch client instead.

**How to verify it**

- The original `FlinkPravegaInputFormatTest.testBatchInput()` is expanded to verify multiple passes of the input format, as well as reading a set of multiple input streams.
- A new `FlinkPravegaInputFormatTest.testBatchInputWithFailure()` is added to verify the behaviour in the presence of failures.

The above tests all fail without the rework in `FlinkPravegaInputFormat`.

EronWright commented on src/main/java/io/pravega/connectors/flink/FlinkPravegaInputFormat.java:
```diff
@@ -76,104 +76,112 @@
             final URI controllerURI,
             final String scope,
             final Set<String> streamNames,
-            final long startTime,
```
This is a breaking change (that I am fine with).

tzulitai replied:
Yes. I also think it is ok, since the batch connector is not yet in any official release, correct?

EronWright commented on src/main/java/io/pravega/connectors/flink/PravegaInputSplit.java:
```diff
@@ -0,0 +1,62 @@
+/**
+ * A {@link PravegaInputSplit} corresponds to a Pravega {@link Segment}.
+ */
+public class PravegaInputSplit implements InputSplit {
```
Is there any requirement that the split implement equality or comparison checks?

tzulitai replied:
I don't think so. Though, it would not hurt to implement that (pretty sure we'll need it eventually for the unit tests anyway). Will add.

tzulitai commented:
@EronWright thanks for the review!
I've addressed your comments. Also, a follow-up issue for writing proper unit tests for the input format was already opened here: https://github.com/pravega/flink-connectors/issues/78.

EronWright commented on src/main/java/io/pravega/connectors/flink/PravegaInputSplit.java:
```diff
@@ -0,0 +1,104 @@
+    public int hashCode() {
+        int result = splitId;
+
+        // Pravega's Segment does not have hashCode implemented
```
I think it does, thanks to the `@Data` lombok annotation.

tzulitai replied:
Ah, I didn't realize that. Thanks, will change.

EronWright commented:
This is great, thanks so much.    If you have a little more time, just make use of `Segment::hashCode()` and we're good to merge.

tzulitai commented:
@EronWright updated 👍
